compiler/rustc_target/src/abi/call/riscv.rs

```diff
@@ -87,7 +87,31 @@ where
             FieldsShape::Primitive => {
                 unreachable!("aggregates can't have `FieldsShape::Primitive`")
             }
-            FieldsShape::Union(_) => {
+            FieldsShape::Union(count) => {
+                if arg_layout.is_transparent() {
+                    let mut non_1zst_elem = None;
+                    for i in 0..count.get() {
+                        let elem_layout = arg_layout.field(cx, i);
+                        if elem_layout.is_1zst() {
+                            continue;
+                        }
+                        // There can only be one.
+                        if non_1zst_elem.is_some() {
+                            unreachable!("two non-1-ZST fields in repr(transparent) union");
+                        }
+                        non_1zst_elem = Some(elem_layout);
+                    }
```

Comment:
Would the `non_1zst_field()` function introduced by #115798 work here? Also I think you will have to recursively unwrap transparent types.

Reply:
This is already recursive, it calls `should_use_fp_conv_helper` on the non-1-ZST field.